datashader/geo.py

```diff
@@ -0,0 +1,472 @@
+
+@ngjit
+def _horn_aspect(data):
+    out = np.zeros_like(data)
+    rows, cols = data.shape
+    for y in range(1, rows-1):
+        for x in range(1, cols-1):
+
+            a = data[y+1, x-1]
+            b = data[y+1, x]
+            c = data[y+1, x+1]
+            d = data[y, x-1]
+            f = data[y, x+1]
+            g = data[y-1, x-1]
+            h = data[y-1, x]
+            i = data[y-1, x+1]
+
+            dz_dx = ((c + 2 * f + i) - (a + 2 * d + g)) / 8
+            dz_dy = ((g + 2 * h + i) - (a + 2 * b + c)) / 8
+
+            aspect = np.arctan2(dz_dy, -dz_dx) * 57.29578  # (180 / pi)
+
+            if aspect < 0:
+                out[y, x] = 90.0 - aspect
+            elif aspect > 90.0:
+                out[y, x] = 360.0 - aspect + 90.0
+            else:
+                out[y, x] = 90.0 - aspect
+
+    return out
+
+
+def aspect(agg):
+    """Returns downward slope direction in compass degrees (0 - 360) with 0 at 12 o'clock.
+
+    Parameters
+    ----------
+    agg : DataArray
+
+    Returns
+    -------
+    data: DataArray
+    """
+
+    if not isinstance(agg, DataArray):
+        raise TypeError("agg must be instance of DataArray")
+
+    return DataArray(_horn_aspect(agg.data),
+                     dims=['y', 'x'],
+                     attrs=agg.attrs)
+
+
+def color_values(agg, color_key, alpha=255):
+
+    def _convert_color(c):
+        r, g, b = rgb(c)
+        return np.array([r, g, b, alpha]).astype(np.uint8).view(np.uint32)[0]
+
+    _converted_colors = {k: _convert_color(v) for k, v in color_key.items()}
+    f = np.vectorize(lambda v: _converted_colors.get(v, 0))
+    return tf.Image(f(agg.data))
+
+
+@ngjit
+def _binary(data, values):
+    out = np.zeros_like(data)
+    rows, cols = data.shape
+    for x in range(0, rows):
+        for y in range(0, cols):
+            if data[y, x] in values:
+                out[y, x] = True
+            else:
+                out[y, x] = False
+    return out
+
+def binary(agg, values):
+    return DataArray(_binary(agg.data, values),
+                     dims=['y', 'x'],
+                     attrs=agg.attrs)
+
+@ngjit
+def _mean(data, excludes):
+    out = np.zeros_like(data)
+    rows, cols = data.shape
+    for y in range(1, rows-1):
+        for x in range(1, cols-1):
+
+            exclude = False
+            for ex in excludes:
+                if data[y,x] == ex:
+                    exclude = True
+                    break
+
+            if not exclude:
+                a,b,c,d,e,f,g,h,i = [data[y-1, x-1], data[y, x-1], data[y+1, x-1],
+                                     data[y-1, x],   data[y, x],   data[y+1, x],
+                                     data[y-1, x+1], data[y, x+1], data[y+1, x+1]]
+                out[y, x] = (a+b+c+d+e+f+g+h+i) / 9
+            else:
+                out[y, x] = data[y, x]
+    return out
+
+def mean(agg, passes=1, excludes=[np.nan]):
+    """
+    Returns Mean filtered array using a 3x3 window sum number of times
```

Comment:
"Returns mean filtered array using a 3x3 window applied in multiple passes"?

Reply:
Yes...revised...